ext/bcmath/libbcmath/src/div.c

```diff
@@ -31,219 +31,428 @@
+		}
+		quot_vector[quot_top_index - i] = n1_vector[n1_top_index - i] / n2_vector;
+		n1_vector[n1_top_index - i] -= n2_vector * quot_vector[quot_top_index - i];
+		n1_vector[n1_top_index - i - 1] += n1_vector[n1_top_index - i] * BC_VECTOR_BOUNDARY_NUM;
+		n1_vector[n1_top_index - i] = 0;
+	}
+	/* last */
+	quot_vector[0] = n1_vector[0] / n2_vector;
+}
 
-	if (digit == 0) {
-		memset(result, 0, size);
-	} else {
-		if (digit == 1) {
-			memcpy(result, num, size);
-		} else {
-			/* Initialize */
-			nptr = (unsigned char *) (num + size - 1);
-			rptr = (unsigned char *) (result + size - 1);
-			carry = 0;
-
-			while (size-- > 0) {
-				value = *nptr-- * digit + carry;
-				*rptr-- = value % BASE;
-				carry = value / BASE;
-			}
+/*
+ * Used when the number of elements in the n2 vector is 2 or more.
+ */
+static inline void bc_standard_div(
+	BC_VECTOR *n1_vector, size_t n1_arr_size,
+	BC_VECTOR *n2_vector, size_t n2_arr_size, size_t n2_len,
+	BC_VECTOR *quot_vector, size_t quot_arr_size
+) {
+	size_t i, j;
+	size_t n1_top_index = n1_arr_size - 1;
+	size_t n2_top_index = n2_arr_size - 1;
+	size_t quot_top_index = quot_arr_size - 1;
+
+	BC_VECTOR div_carry = 0;
+
+	/*
+	 * Consider the error E between the actual quotient and the temporary quotient.
+	 * For example, the quotient of 240 / 121 is 1, but if calculate the quotient only using the high-order digits (24 / 12),
+	 * it will be 2. In this case, 2 is the temporary quotient. Here, the error E is 1.
+	 * Also note that for example 2400000 / 120, there will be 5 divisions.
+	 *
+	 * Another example: Calculating 99900000000 / 10009999 with radix 10000.
+	 * The actual quotient is 9980, but if calculate it using only the high-order digits:
+	 * 999 / 1000 = 0, Numbers that are not divisible are carried forward to the next division.
+	 * 9990000 / 1000 = 9990 (Here, the temporary quotient is 9990 and the error E is 10.)
+	 *
+	 * If calculate the temporary quotient using only one array of n1 and n2, the error E can be larger than 1.
+	 * In other words, in the restoring division, the count of additions for restore increases significantly.
+	 *
+	 * Therefore, in order to keep the error within 1 and to limit the number of additions required for restoration to one
+	 * at most, adjust the number of high-order digits used to calculate the temporary quotient as follows.
+	 * - Adjust the number of digits of n2 used in the calculation to BC_VECTOR_SIZE + 1 digit. The missing digits are
+	 *   filled in from the next array element.
+	 * - Add digits to n1 in the same way as the number of digits adjusted by n2.
+	 *
+	 * e.g. n1 = 123456780000, n2 = 123456789, radix = 10000
+	 * n1_arr = [0, 5678, 1234], n2_arr = [6789, 2345, 1]
+	 * n1_top = 1234, n2_top = 1
+	 * n2_top_tmp = 12345 (+4 digits), n1_top_tmp = 12345678 (+4 digits)
+	 * tmp_quot = n1_top_tmp / n2_top_tmp = 1000, tmp_rem = -9000
+	 * tmp_quot is too large, so tmp_quot--, tmp_rem += n2 (restoring)
+	 * quot = 999, rem = 123447789
```

Comment:
I'm confused by the set-up here, why not just one declaration per line?

Reply:
Fixed